explore/realspace.py cleanup

pkienzle opened the pull request:
Some code cleanup on explore/realspace.py with minor reformatting and removing deprecated functions.

This code is not part of the sasmodels library or the sasview application. Its purpose is to validate analytical models with Monte Carlo sampling on real-space models.

To see options for running the code type:
```sh
$ python explore/realspace.py --help
```

The Pr curve should match the analytic curve but the normalization is wrong so it is shifted. The Iq_avg curve is only valid for radially symmetric samples, so in general it will not match.

lucas-wilkins commented on explore/realspace.py:
```diff
@@ -588,7 +588,7 @@ def spin_weights(in_spin, out_spin):
 
 def orth(A, b_hat): # A = 3 x n, and b_hat unit vector
- return A - np.sum(A*b_hat[:, None], axis=0)[None, :]*b_hat[:, None]    
+    return A - np.sum(A*b_hat[:, None], axis=0)[None, :]*b_hat[:, None]
```
The following goes a beyond the whitespace fixing you did, but:

1) The [:,None] syntax is a bit opaque, maybe b_hat.reshape((3,1))
2) As its used twice, consider declaring it as a variable
3) Consider a more descriptive name and/or comment saying what it does - it's the vector rejection for a an array of vectors right?
4) As the standard formula for this is x - (x dot b_hat) b_hat, maybe code it in terms of np.dot instead.

pkienzle replied:
More generally it would be `b_hat.reshape(-1, 1)`. The form `b_hat[:, None]` is pretty common, though it does something [slightly different](https://stackoverflow.com/questions/28385666/numpy-use-reshape-or-newaxis-to-add-dimensions/28385957#28385957). I find the `b_hat[:, None]` form to more readable in expressions and it is ever so slightly faster in my tests, so I will continue using it where appropriate.

Agreed on replacing sum. Given A is 3 x n and b is 3 x 1, then we want `A - (b (A' b)')' = A - b b' A`, which is conveniently expressed as `A - np.outer(b_hat)@A`. This also happens to be faster than `A - b (b' A)`.

I will change it.

lucas-wilkins replied:
Nobody chooses python because its fast

dehoni replied:
A - np.outer(b, b)@A is also how it was defined in geni.py. I think to remember there we had a discussion about different ways to construct the orthogonal vector fast and supported by numba.

lucas-wilkins reviewed (approved):
All very minor changes, nothing wrong with them.

lucas-wilkins commented on explore/realspace.py:
```diff
@@ -105,23 +110,25 @@ def volume(self):
+        # type: () -> Tuple[float, float, float]
         raise NotImplementedError()
 
     def rotate(self, theta, phi, psi):
```
What coordinate system are these defined in?

pkienzle replied:
dims is the shape bounding box, distance units unspecified.

The rotation angles are a more complicated question since I'm sometimes using Euler angles and other times using Tait-Bryan pseudo-nautical units with (ψ,θ,φ) = (roll, pitch, yaw) around shape axes (c, a, b) corresponding to lab coordinates (z, x, y) with z in the direction of the beam and (x,y) the detector (horizontal, vertical). I didn't write this down in the code since I'm not certain it is correct. I suppose I can add it with a disclaimer.

lucas-wilkins replied:
I always worry when I see Euler angles etc. So easy to get them wrong.

Mainly though, I was fishing for a comment that makes it clear which rotation is applied first (i.e. is it in the "world" or "object" frame)

pkienzle replied:
Yes. I wrote `sasmodels/jitter.py` in order to get it right.

It turns out that my simulation is rotating q space rather than rotating the shape so the rotation code wasn't getting exercised. Which also explains why it can get away with using the "wrong" orientation definition.

There are two orientation systems in sasmodels, one for orienting the shape and another for applying shape distributions. You can see this by running `python -m sasmodels.jitter` setting Δθ=6° and Δφ=10° (less than 5° is treated as zero). This shows the mesh of orientations that are sampled in the integral average. Adjust θ, φ and ψ to show that the mesh is the same regardless of particle orientation. The Euler system of yaw, pitch and roll is better for user control: φ rotates the pattern around in the plane, θ moves the c-axis into the plane, ψ rolls the shape on its axis. Much easier than the previous system of Tait-Bryan coordinates.

The particle tumbling is easier to understand using nautical angles. It isn't ideal since there is pinching at the poles. Compare (large Δθ, small Δφ) to (small Δθ, large Δφ) in `python -m sasmodels.jitter`.  There are map projections that handle this better but we opted for simplicity.

lucas-wilkins commented on explore/realspace.py:
```diff
@@ -105,23 +110,25 @@ def volume(self):
 
     def dims(self):
-        # type: () -> float, float, float
+        # type: () -> Tuple[float, float, float]
```
Tuple needs to be imported

lucas-wilkins commented on explore/realspace.py:
```diff
@@ -105,23 +110,25 @@ def volume(self):
 
     def sample(self, density):
-        # type: (float) -> np.ndarray[N], np.ndarray[N, 3]
+        # type: (float) -> Tuple[np.ndarray[N], np.ndarray[N, 3]]
```
np.ndarray is not a parameterisable class, should be Tuple[np.ndarray, np.ndarray], as much as we might not like it

lucas-wilkins replied:
The downside is that it breaks type inferences

lucas-wilkins commented:
The only things that really require change is import of Tuple, and the removing the parameterisation on ndarray

pkienzle commented on explore/realspace.py:
```diff
@@ -91,11 +93,14 @@ def invert_view(qx, qy, view):
     q = np.vstack((qx.flatten(), qy.flatten(), 0*qx.flatten()))
-    return np.asarray((Rz(-psi)*Ry(-theta)*Rz(-phi))*np.matrix(q))
+    # CRUFT: py3 allows ((Rz(phi) @ Ry(theta) @ Rz(psi)) @ q
+    Rinv = dot(dot(Rz(-psi), Ry(-theta)), Rz(-phi))
```
`a*b*c` expands to `mul(mul(a, b), c)` which for `np.matrix` is equivalent to `dot(dot(a, b), c)`. With the addition of the `@` operator numpy is deprecating the `matrix` class, hence the change. The change itself doesn't warrant a unit test. Whether there should be unit tests on this validation code is a separate question.

I'm using `dot` rather than `@` since sasmodels is still supporting python 2.7 for no particularly good reason.

lucas-wilkins replied:
Core mathematical functions are where you need to have the highest trust, so that's where you want your tests - especially when it's unavoidably hard to read.

lucas-wilkins commented on explore/realspace.py:
```diff
@@ -91,11 +93,14 @@ def invert_view(qx, qy, view):
 
 class Shape:
-    rotation = np.matrix([[1., 0, 0], [0, 1, 0], [0, 0, 1]])
+    rotation = I3
```
Oh yeah, its a class variable. Then it's only ever going to be evaluated once. Definitely no need to define I3 then.

pkienzle commented:
@dehoni I modified the magnetic calculations a bit to use the numpy `@` operator instead of `dot`. I also simplified the code for `orth`. I compared it to a magnetic sphere model and it still seems to produce the same results as the analytic calculation so I probably didn't break anything, but you might want to check the code.

dehoni commented on explore/realspace.py:
```diff
@@ -38,64 +38,68 @@ def Rx(angle):
-    """
+    r"""
     Apply the jitter transform to a set of points.
     Points are stored in a 3 x n numpy matrix, not a numpy array or tuple.
```
These are numpy arrays now as matrix is no longer recommended to use

dehoni commented on explore/realspace.py:
```diff
@@ -38,64 +38,68 @@ def Rx(angle):
+    r"""
+    Return $(q_a, q_b, q_c)$ for the $(\theta, \phi, \psi)$ view angle at
+    detector pixel corresponding to $(q_x, q_y)$.
+    View is in degrees using $z$-$y$-$z$ Euler angles $\phi$-$\theta$-$\psi$.
```
To make it maybe clearer that it is not a mispelling:
using a $z$-$y$-$z$ rotation sequence of the Euler angles ...

dehoni commented on explore/realspace.py:
```diff
@@ -603,16 +606,14 @@ def magnetic_sld(qx, qy, up_theta, up_phi, rho, rho_m):
     M_perp = orth(M,q_hat)
     M_perpP = orth(M_perp, p_hat)
     M_perpP_perpQ = orth(M_perpP, q_hat)
 
-    perpx = np.dot(p_hat, M_perp)
+    perpx = p_hat @ M_perp
     perpy = np.sqrt(np.sum(M_perpP_perpQ**2, axis=0))
-    perpz = np.dot(q_hat, M_perpP)
-    
+    perpz = q_hat @ M_perpP
```
It turned out that this is overly complicated...
Simpler is:
    p_hat = np.array([sin_theta * cos_phi, sin_theta * sin_phi, cos_theta ])
    #two unit vectors spanning up the plane perpendicular to polarisation for SF scattering (avoiding repetitive computation of orthogonal vectors)
    perpy_hat = np.array([-sin_phi, cos_phi, 0 ])
    perpz_hat = np.array([-cos_theta * cos_phi, -cos_theta * sin_phi, sin_theta ])  

    q_hat = np.array([qx, qy, 0]) * q_norm
    M_perp = orth(M, q_hat)

#then the final results are

    perpx = p_hat @ M_perp
    perpy = perpy_hat @ M_perp
    perpz = perpz_hat @ M_perp

pkienzle replied:
Reorganized as:
```python
    # Handle q=0 by setting px = py = 0
    # Note: this is different from kernel_iq, which I(0,0) to 0
    q_norm = 1/sqrt(qx**2 + qy**2) if qx != 0. or qy != 0. else 0.
    q_hat = np.array([qx, qy, 0]) * q_norm

    # perpy_hat and perpz_hat are unit vectors spanning up the plane
    # perpendicular to polarisation for SF scattering (avoiding
    # repetitive computation of orthogonal vectors)
    cos_theta, sin_theta = cos(radians(up_theta)), sin(radians(up_theta))
    cos_phi, sin_phi = cos(radians(up_phi)), sin(radians(up_phi))
    p_hat = np.array([sin_theta * cos_phi, sin_theta * sin_phi, cos_theta])
    perpy_hat = np.array([-sin_phi, cos_phi, 0])
    perpz_hat = np.array([-cos_theta * cos_phi, -cos_theta * sin_phi, sin_theta])

    M_perp = orth(rho_m, q_hat)  # M = rho_m
    perpx = p_hat @ M_perp
    perpy = perpy_hat @ M_perp
    perpz = perpz_hat @ M_perp
```

pkienzle replied:
Is magnetic scattering at |q| = 0 defined? The problem is that lim |q| → 0 magnetic sld depends on the direction q hat.

dehoni replied:
For $q=0$, one would see the demagnetising field of the sample, such that $q_hat = np.sqrt(np.array([0.5, 0.5, 0]))$  for a disc shaped sample that is very thin along the beam.

dehoni replied:
It should never be zero, except the magnetisation direction is perpendicular. Something between 1/3 and 1 is most reasonable and physical motivated.

dehoni commented on explore/realspace.py:
```diff
@@ -636,7 +637,7 @@ def calc_Iq_magnetic(qx, qy, rho, rho_m, points, volume=1.0, view=(0, 0, 0),
     *up_frac_f* is the portion of analyzer neutrons which are spin up.
-    *up_theta* and *up_phi* are the rotation angle of the spin up direction 
+    *up_theta* and *up_phi* are the rotation angle of the spin up direction
     in the detector plane and the inclination from the beam direction (z-axis).
```
Order is reversed: theta is the inclination from the beam direction and phi is the rotation in the detector plance

wpotrzebowski commented:
@dehoni can you check please if it is ok now?

wpotrzebowski commented:
@dehoni will take a quick look and we hopefully can merge it

dehoni commented:
Sorry, my bad, I did not find the box to approve the changes...
